File: java/src/main/java/ai/buble/sdk/chat/GeminiService.java
package ai.buble.sdk.chat;

import ai.buble.sdk.RequestOptions;
import ai.buble.sdk.http.BubleHttpClient;
import ai.buble.sdk.streaming.BubleStream;
import ai.buble.sdk.streaming.StreamProtocol;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;

import java.net.http.HttpResponse;
import java.util.Map;

/**
 * Gemini-compatible content generation methods.
 */
public final class GeminiService {
    private static final TypeReference<Map<String, JsonNode>> CHAT_RESPONSE =
            new TypeReference<Map<String, JsonNode>>() {};

    private final BubleHttpClient http;

    public GeminiService(BubleHttpClient http) {
        this.http = http;
    }

    public Map<String, JsonNode> generateContent(String model, Map<String, Object> body) {
        String path = "/api/v1beta/models/" + BubleHttpClient.encodeModelPath(model) + ":generateContent";
        return http.post(path, ChatCompletionsService.copy(body), CHAT_RESPONSE);
    }

    public BubleStream streamGenerateContent(String model, Map<String, Object> body) {
        String path = "/api/v1beta/models/" + BubleHttpClient.encodeModelPath(model) + ":streamGenerateContent";
        HttpResponse<java.io.InputStream> response = http.stream(path, ChatCompletionsService.copy(body), RequestOptions.none());
        return new BubleStream(response.body(), http.getObjectMapper(), StreamProtocol.GEMINI);
    }
}

File: php/src/Chat/GeminiService.php
<?php

declare(strict_types=1);

namespace Buble\Chat;

use Buble\Http\TransportInterface;
use Buble\Streaming\SseEvent;
use Buble\Streaming\StreamText;

final class GeminiService
{
    public function __construct(private readonly TransportInterface $transport)
    {
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    public function generateContent(string $model, array $body): array
    {
        return $this->transport->request(
            'POST',
            '/api/v1beta/models/' . $this->encodeModelPath($model) . ':generateContent',
            $body,
        );
    }

    /**
     * @param array<string, mixed> $body
     * @return \Generator<int, SseEvent>
     */
    public function streamGenerateContent(string $model, array $body): \Generator
    {
        yield from $this->transport->stream(
            '/api/v1beta/models/' . $this->encodeModelPath($model) . ':streamGenerateContent',
            $body,
        );
    }

    /**
     * @param array<string, mixed> $body
     * @return \Generator<int, string>
     */
    public function streamText(string $model, array $body): \Generator
    {
        yield from StreamText::fromEvents($this->streamGenerateContent($model, $body), 'gemini');
    }

    private function encodeModelPath(string $model): string
    {
        return implode('/', array_map('rawurlencode', explode('/', $model)));
    }
}
